src/ncp/async_task.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Copyright (c) 2024, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for chained async task.
+ *   The utility class is used to support the usage of a `Then`-style chained async operations.
+ */
+
+#ifndef OTBR_AGENT_ASYNC_TASK_HPP_
+#define OTBR_AGENT_ASYNC_TASK_HPP_
+
+#include <functional>
+#include <memory>
+
+#include <openthread/error.h>
+
+namespace otbr {
+namespace Ncp {
+
+class AsyncTask;
+using AsyncTaskPtr = std::shared_ptr<AsyncTask>;
+
+class AsyncTask
+{
+public:
+    using ThenHandler  = std::function<void(AsyncTaskPtr)>;
+    using ErrorHandler = std::function<void(otError, const std::string &)>;
+
+    /**
+     * Constructor.
+     *
+     * @param[in]  The error handler called when the result is not OT_ERROR_NONE;
```

Comment:
I think if all the chained tasks are finished successfully we will call `mErrorHandler` with `OT_ERROR_NONE` to indicate this.
```c++
void AsyncTask::SetResult(otError aError, const std::string &aErrorInfo)
{
    if (aError == OT_ERROR_NONE)
    {
        if (mThen)
        {
            mThen(std::move(mNext));
        }
        else if (mErrorHandler)
        {
            // The Final case
            mErrorHandler(aError, aErrorInfo);
        }
    }
```

How about renaming it to `ResultHandler` (it is not just used for reporting errors but successful completion of executing all the chained tasks?)

Reply:
Have renamed it.

Reply:
Thanks.